File: bot/src/gpt/tool/function/SpeakFunction.ts
import AssistantFunction, {AssistantObject, InvocationContext} from "./AssistantFunction";
import {VkMessage} from "../../../service/VkMessagesService";
import {Context} from "../../../Context";

export default class SpeakFunction implements AssistantFunction {
    constructor(private context: Context) {}

    async call(args: any, message: VkMessage, invocationContext: InvocationContext): Promise<string> {
        const text = args['text'];
        const chatSettings = await this.context.chatSettingsService.getSettingsOrCreateDefault(message.peerId);
        await this.context.vkMessagesService.indicateActivity(message.peerId, "audiomessage");
        const audio = await this.context.audioService.generateSpeech(
            text,
            "tts-1-hd",
            chatSettings.ttsVoice,
            chatSettings.ttsSpeed
        );
        const attachment = await this.context.vkMessagesService.uploadVoiceMessage(message.peerId, audio);
        invocationContext.addAttachment(attachment);
        return "Audio message is attached";
    }

    getDescription(): string {
        return "Attach an audio message. " +
            "Use this to speak with voice. " +
            "Prefer to speak with voice with people who also speak with voice messages. " +
            "Audio messages can't have other attachments, don't mix with 'draw'.";
    }

    getName(): string {
        return "speak";
    }

    getParameters(): AssistantObject | null {
        return {
            type: "object",
            properties: {
                text: {
                    type: "string",
                    description: "Text to speak"
                }
            },
            required: ["text"]
        };
    }

}

File: backend/src/main/java/space/davids_digital/sweetie/gpt/tool/function/SpeakFunction.kt
package space.davids_digital.sweetie.gpt.tool.function

import com.vk.api.sdk.objects.messages.SetActivityType
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.stereotype.Component
import space.davids_digital.sweetie.gpt.InvocationContext
import space.davids_digital.sweetie.gpt.tool.function.parameter.Description
import space.davids_digital.sweetie.integration.openai.OpenAiService
import space.davids_digital.sweetie.integration.vk.VkMessageService
import space.davids_digital.sweetie.model.VkMessageModel
import space.davids_digital.sweetie.service.ChatSettingsService

@Component
class SpeakFunction(
    private val chatSettingsService: ChatSettingsService,
    private val vkMessageService: VkMessageService,
    private val openAiService: OpenAiService
): AssistantFunction<SpeakFunction.SpeakParameters> {
    data class SpeakParameters (
        @Description("Text to speak")
        val text: String
    )

    override fun getName() = "speak"
    override fun getParametersClass() = SpeakParameters::class
    override fun getDescription() = """
        Attach an audio message.
        Use this to speak with voice.
        Prefer to speak with voice with people who also speak with voice messages.
        Audio messages can't have other attachments, don't mix with 'draw'.
    """.trimIndent()

    override suspend fun call(
        parameters: SpeakParameters,
        message: VkMessageModel,
        invocationContext: InvocationContext
    ): String {
        val chatSettings = withContext(Dispatchers.IO) {
            chatSettingsService.getOrCreateDefault(message.peerId)
        }
        vkMessageService.indicateActivity(message.peerId, SetActivityType.AUDIOMESSAGE)
        val audio = openAiService.speech(parameters.text, "tts-1-hd", chatSettings.ttsVoice, chatSettings.ttsSpeed)
        val attachment = vkMessageService.uploadVoiceMessageAttachment(message.peerId, audio)
        invocationContext.addAttachment(attachment)
        return "Audio message is attached."
    }
}